data/src/main/java/org/cryptomator/data/cloud/pcloud/PCloudImpl.java

```diff
@@ -0,0 +1,418 @@
+			throws IOException, BackendException {
+		if (!replace && exists(file)) {
+			throw new CloudNodeAlreadyExistsException("CloudNode already exists and replace is false");
+		}
+
+		if (file.getParent().getId() == null) {
+			throw new NoSuchCloudFileException(String.format("The parent folder of %s doesn't have a folderId. The file would remain in root folder", file.getPath()));
+		}
+
+		progressAware.onProgress(Progress.started(UploadState.upload(file)));
+		UploadOptions uploadOptions = UploadOptions.DEFAULT;
+		if (file.getId() != null && replace) {
+			uploadOptions = UploadOptions.OVERRIDE_FILE;
+		}
+
+		RemoteFile uploadedFile = uploadFile(file, data, progressAware, uploadOptions, size);
+
+		progressAware.onProgress(Progress.completed(UploadState.upload(file)));
+
+		return idCache.cache(PCloudNodeFactory.file(file.getParent(), uploadedFile));
+	}
+
+	private RemoteFile uploadFile(final PCloudFile file, DataSource data, final ProgressAware<UploadState> progressAware, UploadOptions uploadOptions, final long size) //
+			throws IOException, BackendException {
+			ProgressListener listener = (done, total) -> progressAware.onProgress( //
+					progress(UploadState.upload(file)) //
+							.between(0) //
+							.and(size) //
+							.withValue(done));
+
+		com.pcloud.sdk.DataSource pCloudDataSource = new com.pcloud.sdk.DataSource() {
+			@Override
+			public long contentLength() {
+				return data.size(context).get();
+			}
+
+			@Override
+			public void writeTo(BufferedSink sink) throws IOException {
+				try (Source source = Okio.source(data.open(context))) {
+					sink.writeAll(source);
+				}
+			}
+		};
+
+		Long parentFolderId = file.getParent().getId();
+		if (parentFolderId == null) {
+			parentFolderId = idCache.get(file.getParent().getPath()).getId();
+		}
+
+		String filename = file.getName();
+		String encodedFilename = URLEncoder.encode(filename, UTF_8);
+
+		try {
+			RemoteFile newFile = client() //
+					.createFile(parentFolderId, encodedFilename, pCloudDataSource, new Date(), listener, uploadOptions) //
+					.execute();
+			if (!filename.equals(encodedFilename)) {
+				return client().renameFile(newFile.fileId(), filename).execute();
+			}
+			return newFile;
+		} catch (ApiError ex) {
+			handleApiError(ex);
+			throw new FatalBackendException(ex);
+		}
+	}
+
+	public void read(PCloudFile file, OutputStream data, final ProgressAware<DownloadState> progressAware) throws IOException, BackendException {
+		progressAware.onProgress(Progress.started(DownloadState.download(file)));
+
+		Long fileId = file.getId();
+		if (fileId == null) {
+				PCloudIdCache.NodeInfo nodeInfo = idCache.get(file.getPath());
+				if (nodeInfo != null) {
+					fileId = nodeInfo.getId();
+				} else {
+					Optional<RemoteEntry> remoteEntryOptional = findEntry(file.getParent().getId(), file.getName(), false);
+					if (remoteEntryOptional.isPresent()) {
+						fileId = remoteEntryOptional.get().asFile().fileId();
+					}
```

Comment:
Doesn't make it more readable but if the file really not exists, we need to throw a `NoSuchCloudFileException`:

```suggestion
					} else {
						throw new NoSuchCloudFileException(file.getName());
					}
```

This prevents `createFileLink()` throwing a `Nullpointer` and allows the UI to display the NotFound-dialog.